asserts/tool/gpgkeypairmgr.go

```diff
@@ -0,0 +1,136 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package tool
+
+import (
+	"bytes"
+	"encoding/hex"
+	"fmt"
+	"io"
+	"os/exec"
+	"strings"
+
+	"golang.org/x/crypto/openpgp/packet"
+
+	"github.com/snapcore/snapd/asserts"
+)
+
+type gpgKeypairManager struct {
+	homedir string
+}
+
+func (gkm *gpgKeypairManager) gpg(input []byte, args ...string) ([]byte, error) {
+	general := []string{"-q"}
+	if gkm.homedir != "" {
+		general = append([]string{"--homedir", gkm.homedir}, general...)
+	}
+	allArgs := append(general, args...)
+
+	cmd := exec.Command("gpg", allArgs...)
+	var outBuf bytes.Buffer
+	var errBuf bytes.Buffer
+
+	if len(input) != 0 {
+		cmd.Stdin = bytes.NewBuffer(input)
+	}
+
+	cmd.Stdout = &outBuf
+	cmd.Stderr = &errBuf
+
+	if err := cmd.Run(); err != nil {
+		return nil, fmt.Errorf("gpg %q failed: %v (%q)", strings.Join(args, " "), err, errBuf.Bytes())
+	}
+
+	return outBuf.Bytes(), nil
+}
+
+// NewGPGKeypairManager creates a new key pair manager backed by a local GnuPG setup using the given GPG homedir,
+// and asking GPG to fallback "~/.gnupg" to default if
+// empty. Importing keys through the keypair manager interface is not
+// supported. Main purpose is allowing signing using keys from a GPG setup.
+func NewGPGKeypairManager(homedir string) asserts.KeypairManager {
+	return &gpgKeypairManager{
+		homedir: homedir,
+	}
+}
+
+func (gkm *gpgKeypairManager) Put(authorityID string, privKey asserts.PrivateKey) error {
+	// NOTE: we don't need this initially at least and this keypair mgr is not for general arbitrary usage
+	return fmt.Errorf("importing key in a GPG keypair manager is not supported")
+}
+
+func (gkm *gpgKeypairManager) Get(authorityID, keyID string) (asserts.PrivateKey, error) {
+	out, err := gkm.gpg(nil, "--batch", "--export", "--export-options", "export-minimal,export-clean,no-export-attributes", "0x"+keyID)
+	if err != nil {
+		return nil, err
+	}
+	if len(out) == 0 {
+		return nil, fmt.Errorf("no matching key pair found")
+	}
+
+	var pubKey *packet.PublicKey
+	rd := packet.NewReader(bytes.NewBuffer(out))
+	for {
+		pkt, err := rd.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, fmt.Errorf("cannot read public part of key pair: %v", err)
+		}
+		cand, ok := pkt.(*packet.PublicKey)
+		if ok {
+			if cand.IsSubkey {
+				continue
+			}
+			if pubKey != nil {
+				return nil, fmt.Errorf("cannot find exactly one key pair with key id %q, found many", keyID)
+			}
+			pubKey = cand
+		}
+	}
+
+	if pubKey == nil {
+		return nil, fmt.Errorf("cannot read public part of key pair: unexpectedly missing")
+
+	}
+
+	sign := func(content []byte, cfg *packet.Config) (*packet.Signature, error) {
```

Comment:
Reading this also gives me a feeling that something might be simplified on this design.